pkg/filters/redirector/redirector_test.go

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package redirector
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/protocols/httpprot"
+	"github.com/stretchr/testify/assert"
+)
+
+func getSpec(match string, part string, replace string, code int) *Spec {
+	return &Spec{
+		Match:       match,
+		MatchPart:   part,
+		Replacement: replace,
+		StatusCode:  code,
+	}
+}
+
+func TestRedirector(t *testing.T) {
+	assert := assert.New(t)
+
+	type match struct {
+		reqURL       string
+		expectedURL  string
+		expectedCode int
+		expectedBody string
+	}
+
+	getMatch := func(req string, expected string, code int, body string) match {
+		return match{
+			reqURL:       req,
+			expectedURL:  expected,
+			expectedCode: code,
+			expectedBody: body,
+		}
+	}
+
+	type testCase struct {
+		spec    *Spec
+		matches []match
+	}
+
+	getMsg := func(caseId int, matchId int) string {
+		return fmt.Sprintf("case %d match %d failed.", caseId, matchId)
+	}
+
+	// test different spec configurations
+	for i, t := range []testCase{
+		{
+			spec: getSpec("(.*)", "", "$1", 0), // use default, uri, 301
+			matches: []match{
+				getMatch("http://a.com:8080/foo/bar?baz=qux", "/foo/bar?baz=qux", 301, "Moved Permanently"),
+			},
+		},
+		{
+			spec: getSpec("(.*)", "uri", "$1", 301), // uri, 301
+			matches: []match{
+				getMatch("http://a.com:8080/foo/bar?baz=qux", "/foo/bar?baz=qux", 301, "Moved Permanently"),
+			},
+		},
+		{
+			spec: getSpec("(.*)", "full", "$1", 302), // full, 302
+			matches: []match{
+				getMatch("http://a.com:8080/foo/bar?baz=qux", "http://a.com:8080/foo/bar?baz=qux", 302, "Found"),
+			},
+		},
+		{
+			spec: getSpec("(.*)", "path", "$1", 302), // path, 302
+			matches: []match{
+				getMatch("http://a.com:8080/foo/bar?baz=qux", "/foo/bar", 302, "Found"),
+			},
+		},
+	} {
+		r := &Redirector{spec: t.spec}
+		r.Init()
+		for j, m := range t.matches {
+			msg := getMsg(i, j)
+			req, err := http.NewRequest(http.MethodGet, m.reqURL, nil)
+			assert.Nil(err, msg)
+			httpReq, err := httpprot.NewRequest(req)
+			assert.Nil(err, msg)
+
+			ctx := context.New(nil)
+			ctx.SetInputRequest(httpReq)
+			r.Handle(ctx)
+
+			resp := ctx.GetOutputResponse().(*httpprot.Response)
+			assert.Equal(m.expectedURL, resp.Header().Get("Location"), msg)
+			assert.Equal(m.expectedCode, resp.StatusCode(), msg)
+			assert.Equal(m.expectedBody, string(resp.RawPayload()), msg)
+		}
+	}
+
+	// test invalid spec
+	for i, t := range []*Spec{
+		getSpec("(.*)", "all", "$1", 0),   // invalid match part
+		getSpec("(.*)", "uri", "$1", 200), // invalid status code
+		getSpec("(+)", "uri", "$1", 301),  // invalid regex
+	} {
+		msg := fmt.Sprintf("case %d failed.", i)
+		r := &Redirector{spec: t}
+		assert.Panics(func() { r.Init() }, msg)
+	}
+
+	// test complicated regex
+	for i, t := range []testCase{
```

Comment:
I'd like to add the following classical test cases:

- **Case 1: URI Prefix Redirect**
   -  Spec: matchPart: `uri`, match `^(.*)$`, replacement `/prefix$1`
   -  input: `https://example.com/path/to/api/?key1=123&key2=456`
   -  output: `https://example.com/prefix/path/to/api/?key1=123&key2=456`

- **Case 2: Domain Redirect**
   -  Spec: matchPart: `url`, match `(^.*\/\/)([^\/]*)(.*$)`, replacement `$1my.com/$3`
   -  input: `https://example.com/path/to/api/?key1=123&key2=456`
   -  output: `https://my.com/path/to/api/?key1=123&key2=456`

- **Case 3: Path Redirect**
   -  Spec: matchPart: `path`, match `(user)\.php\?id=(\d*)`, replacement `/api/$1/$2`
   -  input: `https://example.com/path/to/user.php?id=123`
   -  output: `https://example.com/api/user/123`

Reply:
All updated! Please check